### PR TITLE
Finish records support

### PR DESCRIPTION
**Pull request by twistedsquare**
After https://github.com/typetools/checker-framework/pull/4802 there were two issues left to complete:

 * Allow method annotations on compact constructors in stubs
 * Support automatic transfer from declaration annotations on record components in stubs to the corresponding field, method, and compact canonical constructor parameter.  (Outside stubs this is done automatically by javac.)

This PR (in combination with https://github.com/typetools/stubparser/pull/72) adds support for both these items.  I tried using some existing checkers for the testing but none have straightforward support for annotations on fields, methods and parameters.  So I added my own simple test-only checker "DisbarUse" that gives an error when the element is used, in order to test this functionality.

I think with this, support for records is complete in all aspects, unless I've overlooked something.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java [hunk 1]
```diff
@@ -343,16 +343,44 @@ public RecordStub(LinkedHashMap<String, RecordComponentStub> componentsByName) {
...
+    }
+
+    /**
+     * Get the record component annotations that are applicable to the given element kind.
+     *
+     * @param elementKind the element kind to apply to (e.g., FIELD, METHOD)
+     * @return the set of annotations from the component that apply
+     */
+    public Set<AnnotationMirror> getAnnotationsForTarget(ElementKind elementKind) {
+      HashSet<AnnotationMirror> filtered = new HashSet<>();
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:374]: This use of `HashSet` makes this method nondeterministic.  Below, `AnnotationUtils.createAnnotationSet` returns a deterministic set (in particular, a `TreeSet`), and I think retaining that determinism would be best.

File: checker/src/test/java/org/checkerframework/checker/testchecker/disbaruse/DisbarUseVisitor.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    return super.visitNewClass(node, p);
+  }
+
+  @Override
+  public Void visitIdentifier(IdentifierTree node, Void p) {
+    MemberSelectTree enclosingMemberSel = enclosingMemberSelect();
+    MemberSelectTree[] memberSelectTrees =
+        enclosingMemberSel == null
+            ? new MemberSelectTree[] {null}
+            : new MemberSelectTree[] {enclosingMemberSel, null};
```
[the remark this answers is not in the document]
Comment (twistedsquare) [reply checker/src/test/java/org/checkerframework/checker/testchecker/disbaruse/DisbarUseVisitor.java:54]: Yes, done.

File: checker/src/test/java/org/checkerframework/checker/testchecker/disbaruse/DisbarUseVisitor.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      final Element elem;
+      if (memberSel == null) {
+        tree = node;
+        elem = TreeUtils.elementFromUse(node);
+      } else {
+        tree = memberSel;
+        elem = TreeUtils.elementFromUse(memberSel);
+      }
+
+      if (elem == null || (!elem.getKind().isField() && elem.getKind() != ElementKind.PARAMETER)) {
```
[the remark this answers is not in the document]
Comment (twistedsquare) [reply checker/src/test/java/org/checkerframework/checker/testchecker/disbaruse/DisbarUseVisitor.java:67]: The if condition was confusing; it only warns on fields and parameters.  I've refactored to be clearer and added a comment.

File: checker/src/test/java/org/checkerframework/checker/testchecker/disbaruse/DisbarUseChecker.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package org.checkerframework.checker.testchecker.disbaruse;
+
+import org.checkerframework.common.basetype.BaseTypeChecker;
+
+/**
+ * A checker that issues a "disbar.use" error at any use of an expression whose type is
```
[the remark this answers is not in the document]
Comment (twistedsquare) [reply checker/src/test/java/org/checkerframework/checker/testchecker/disbaruse/DisbarUseChecker.java:6]: I've added a comment; it issues an error for parameters, methods and fields.

File: framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java (hunk 1 above)
Comment (twistedsquare) [reply framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:374]: Done.

**Comment by twistedsquare** [comment]
All the comments have been addressed so I think it is ready to merge.

**Review by mernst — APPROVED** [review]

**Review by smillst — APPROVED** [review]

**Comment by mernst** [comment]
Thanks again for the records feature!

**Pull request merged**